_tour/pattern-matching.md

```diff
@@ -106,40 +163,103 @@ println(showImportantNotification(importantEmail, importantPeopleInfo)) // print
+val importantEmail = Email("jenny@gmail.com", "Drinks tonight?", "I'm free after 5!")
+val importantSms = SMS("867-5309", "I'm here! Where are you?")
+
+println(showImportantNotification(someSms, importantPeopleInfo)) // prints You got an SMS from 123-4567! Message: Are you there?
+println(showImportantNotification(someVoiceRecording, importantPeopleInfo)) // prints You received a Voice Recording from Tom! Click the link to hear it: voicerecording.org/id/123
+println(showImportantNotification(importantEmail, importantPeopleInfo)) // prints You got an email from special someone!
+
+println(showImportantNotification(importantSms, importantPeopleInfo)) // prints You got an SMS from special someone!
+```
+{% endtab %}
+{% endtabs %}
 
 In the `case Email(sender, _, _) if importantPeopleInfo.contains(sender)`, the pattern is matched only if the `sender` is in the list of important people.
 
 ## Matching on type only
 You can match on the type like so:
+{% tabs pattern-matching-6 class=tabs-scala-version %}
+{% tab 'Scala 2' for=pattern-matching-6 %}
 ```scala mdoc
-abstract class Device
+sealed trait Device
 case class Phone(model: String) extends Device {
   def screenOff = "Turning screen off"
 }
 case class Computer(model: String) extends Device {
   def screenSaverOn = "Turning screen saver on..."
 }
 
-def goIdle(device: Device) = device match {
+def goIdle(device: Device): String = device match {
   case p: Phone => p.screenOff
   case c: Computer => c.screenSaverOn
 }
 ```
+{% endtab %}
+{% tab 'Scala 3' for=pattern-matching-6 %}
+```scala
+sealed trait Device
+case class Phone(model: String) extends Device:
+  def screenOff = "Turning screen off"
+
+case class Computer(model: String) extends Device:
+  def screenSaverOn = "Turning screen saver on..."
+
+
+def goIdle(device: Device): String = device match
+  case p: Phone => p.screenOff
+  case c: Computer => c.screenSaverOn
+```
+{% endtab %}
+{% endtabs %}
+
 `def goIdle` has a different behavior depending on the type of `Device`. This is useful when the case needs to call a method on the pattern. It is a convention to use the first letter of the type as the case identifier (`p` and `c` in this case).
 
 ## Sealed classes
 Traits and classes can be marked `sealed` which means all subtypes must be declared in the same file. This assures that all subtypes are known.
 
+{% tabs pattern-matching-7 class=tabs-scala-version %}
+{% tab 'Scala 2' for=pattern-matching-7 %}
 ```scala mdoc
-sealed abstract class Furniture
-case class Couch() extends Furniture
-case class Chair() extends Furniture
+sealed trait Furniture
+case object Couch extends Furniture
+case object Chair extends Furniture
 
 def findPlaceToSit(piece: Furniture): String = piece match {
   case a: Couch => "Lie on the couch"
   case b: Chair => "Sit on the chair"
 }
 ```
+{% endtab %}
+{% tab 'Scala 3' for=pattern-matching-7 %}
+```scala
+sealed trait Furniture
```

Comment:
enum?

Reply:
Yes probably we should use an enum here. The problem is that they have not yet been introduced in the Tour of Scala.

Reply:
Do you mean type parameter or value parameter?
Do you think I should do that in this PR?

Reply:
Here is my suggestion for both:
```scala
sealed abstract class Furniture:
  def descriptor: String

case class Couch(descriptor: String) extends Furniture
case class Chair(descriptor: String) extends Furniture

def findPlaceToSit(piece: Furniture): String = piece match
  case Couch(descriptor) => "Lie on the " + descriptor + " couch"
  case Chair(descriptor) => "Sit on the " + descriptor + " chair"
```

Reply:
Sounds great to me!